src/vm/gchelpers.cpp

```diff
@@ -1467,7 +1465,7 @@ extern "C" HCIMPL2_RAW(VOID, JIT_WriteBarrier, Object **dst, Object *ref)
 #ifdef FEATURE_COUNT_GC_WRITE_BARRIERS
         UncheckedAfterRefInEphemFilter++;
 #endif
-        BYTE* pCardByte = (BYTE *)VolatileLoadWithoutBarrier(&g_card_table) + card_byte((BYTE *)dst);
+        BYTE* pCardByte = (BYTE *)g_card_table + card_byte((BYTE *)dst);
```

Comment:
What make it safe to delete all these `VolatileLoadWithoutBarrier`?

Reply:
For the correctness we need to make sure that g_card_table (and heap bounds if this was a checked barrier) are read after (in program order) than we obtained dst. I do not think this barrier was helping that.

---
In reply to: [293091841](https://github.com/dotnet/coreclr/pull/25130#discussion_r293091841) [](ancestors = 293091841)

Reply:
I.E. - we could read `g_card_table` into a temp at very beginning of this method and later use it and nothing will fail. 
I tried such change in the checked barrier, just to be sure, since I suspected that "WithoutBarrier" was not strong enough on ARM64 and we were reading ahead of bounds. Turned out reading g_card_table ahead of bounds is just fine since we should not be seeing managed objects/locations that are newer than the either the table or the bounds.

---
In reply to: [293097681](https://github.com/dotnet/coreclr/pull/25130#discussion_r293097681) [](ancestors = 293097681,293091841)

Reply:
Looking at this the other way - if we were concerned about getting a card table that is older than the object/destination at hand, we should be concerned about getting old heap bounds too. Evidently we were reading heap bounds with ordinary reads.

---
In reply to: [293100015](https://github.com/dotnet/coreclr/pull/25130#discussion_r293100015) [](ancestors = 293100015,293097681,293091841)

Reply:
GC will reallocate the card table as necessary. It is super rare but it does happen. Also, the card table reallocation is done without suspending the runtime. 

When `g_card_table` is read before checking the bounds in checked barrier, you can end up with pointer to the old small card table, new larger bounds that match the reallocated card table, and a pointer that is covered by the new card table, but covered by the old card table. The net result is that a random somewhere gets flipped to 0xFF once in a blue moon. I do not see anything that prevents this in your change.

BTW: We had this bug for a long time and I have spent like one month of my life tracing this down. It was bug http://vstfdevdiv:8080/DevDiv2/DevDiv/_workitems/edit/346765

Reply:
also I'd like to point out that with the FEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP define it's more rare that we'd call resize when EE is not suspended 'cause if BGC is enabled (and by default it is) we always suspend the EE when we call resize.

Reply:
Besides, if you notice the ` *dst = ref;`  is not a release store. That is ok because this is called only by interlocked operations (although I am not sure that unchecked one is used at all). Anyways - interlocked callers could be considered as one more guarantee that reads of globals will never happen before reading the object or the destination.

I think we are fine with removing these `WithoutBarrier`.

---
In reply to: [293109065](https://github.com/dotnet/coreclr/pull/25130#discussion_r293109065) [](ancestors = 293109065,293100015,293097681,293091841)

Reply:
>  if we were concerned about getting a card table that is older than the object/destination at hand, we should be concerned about getting old heap bounds too. Evidently we were reading heap bounds with ordinary reads.

The assumption is that the compiler won't reorder memory loads over `VolatileLoadWithoutBarrier`.

> I think we are fine with removing these WithoutBarrier.

I doubt it.

Reply:
Yes. I tried reproducing the original issue with BGC disabled, if that what you are suggesting. It did not help with the repro. It should have, but did not in the time I spent trying.

Setting a shorter segment is a good idea. I did not try that. 

---
In reply to: [293115314](https://github.com/dotnet/coreclr/pull/25130#discussion_r293115314) [](ancestors = 293115314)

Reply:
Talked to @jkotas@microsoft.com - yes. the problem is that while heap space cannot encroach into formerly stack space (it is one or another) it can surround it. That is the case where we may have to mark a write into stack space. In particular it results in a scenario where legit "not markable" write becomes "markable" once the new bounds are established. I basically assumed it is either one or another.

The inconvenient consequence is that since bounds and the table are not updated atomically, access to them must be ordered. - We do need a compiler fence on the barrier side and one more process-wide fence on the update side. (the process-wide fences are needed only on weak architectures though). 


---
In reply to: [293128578](https://github.com/dotnet/coreclr/pull/25130#discussion_r293128578) [](ancestors = 293128578)

Reply:
I wonder if storing card table size _on_ the card table could solve some of these troubles. Since every arch that we support respects indirection dependency, we could use fewer fences.
I am not sure I am ready to make a change like this in this PR though. 

---
In reply to: [293168681](https://github.com/dotnet/coreclr/pull/25130#discussion_r293168681) [](ancestors = 293168681,293128578)

Reply:
Tried with shorter segment (and with BGC disabled), but it did not repro. Also tried WKS and server GC.
I suspect on a single machine it may take days to appear. It also could be that "more stress" is not necessarily something that would make the failure to surface. - If it happens because of stale data in caches, too much churn could possibly reduce the chances.

---
In reply to: [293131703](https://github.com/dotnet/coreclr/pull/25130#discussion_r293131703) [](ancestors = 293131703,293115314)

Reply:
I have also done some digging through the bug hits. It looks like it was appearing in test passes 1-2 times a week and we did not have hits since 6/7. 
Being so rare, I really doubt this is reasonably reproducible.

---
In reply to: [293918916](https://github.com/dotnet/coreclr/pull/25130#discussion_r293918916) [](ancestors = 293918916,293131703,293115314)

Reply:
I think, if changes look reasonable, perhaps we should merge the PR. And if the bug still comes back, it is something else - we know that `FlushProcessWriteBuffers` is not reliable for example.

The idea is that the causes for the failure may have compounding effect, so by removing some, we hopefully may make the issue so rare that it is effectively fixed.

---
In reply to: [293967793](https://github.com/dotnet/coreclr/pull/25130#discussion_r293967793) [](ancestors = 293967793,293918916,293131703,293115314)

Reply:
I have run GC reliability tests overnight locally on Linux/ARM64. No failures. That was with default settings. Tried /w no concurrent GC and segment size set to 1K for couple hours. It made no difference.
If the failure is more likely to happen at start up, running test scenarios in a loop may not be a good way to catch it. 

---
In reply to: [293969737](https://github.com/dotnet/coreclr/pull/25130#discussion_r293969737) [](ancestors = 293969737,293967793,293918916,293131703,293115314)

Reply:
There were 2 new hits in lab runs though. After 10 days of no hits.

---
In reply to: [295432752](https://github.com/dotnet/coreclr/pull/25130#discussion_r295432752) [](ancestors = 295432752,293969737,293967793,293918916,293131703,293115314)